File: web/customer/src/api/notifications.ts
import { authenticatedFetch } from './client';

// START_MODULE_CONTRACT
//   PURPOSE: Customer notification feed REST client with runtime shape checks.
//   SCOPE:   NotificationChannel/Type/Status unions, NotificationFeedItem,
//            NotificationFeedResponse, NotificationApiError,
//            listNotifications.
//   DEPENDS: M-CORE-API (HTTP /api/v1/profile/notifications),
//            ./client (authenticatedFetch).
//   LINKS:   docs/development-plan.xml M-WEB-CUSTOMER, PDD §5.2 notifications;
//            INV-013 (own-user scoped feed, no user_id/PII projection).
//   ROLE:    RUNTIME
//   MAP_MODE: EXPORTS
// END_MODULE_CONTRACT
//
// START_MODULE_MAP
//   NotificationChannel      - in_app | sms
//   NotificationType         - order_status_change | otp
//   NotificationStatus       - pending | sent | failed
//   NotificationFeedItem     - one customer-visible notification row
//   NotificationFeedResponse - paginated wrapper
//   NotificationApiError     - Error subclass with HTTP status + detail
//   listNotifications        - GET /profile/notifications?page&per_page
// END_MODULE_MAP

export type NotificationChannel = 'in_app' | 'sms';
export type NotificationType = 'order_status_change' | 'otp';
export type NotificationStatus = 'pending' | 'sent' | 'failed';

export interface NotificationFeedItem {
  id: string;
  order_id: string | null;
  channel: NotificationChannel;
  type: NotificationType;
  status: NotificationStatus;
  message_ru: string;
  message_en: string;
  sent_at: string | null;
  created_at: string;
}

export interface NotificationFeedResponse {
  notifications: NotificationFeedItem[];
  total_count: number;
  page: number;
  per_page: number;
}

// START_CONTRACT: NotificationApiError
//   PURPOSE: Carry HTTP status + detail for notification-feed UI failures.
//   INPUTS:  status: number
//            detail?: string
//   OUTPUTS: NotificationApiError instance.
//   SIDE_EFFECTS: none.
// END_CONTRACT: NotificationApiError
export class NotificationApiError extends Error {
  constructor(
    public status: number,
    public detail?: string,
  ) {
    super(detail ?? `HTTP ${status}`);
    this.name = 'NotificationApiError';
  }
}

async function parseError(res: Response): Promise<NotificationApiError> {
  let detail: string | undefined;
  try {
    const body = (await res.json()) as { detail?: string };
    detail = body?.detail;
  } catch {
    // non-json body
  }
  return new NotificationApiError(res.status, detail);
}

function isNotificationItem(value: unknown): value is NotificationFeedItem {
  if (!value || typeof value !== 'object') return false;
  const row = value as NotificationFeedItem;
  return (
    typeof row.id === 'string' &&
    (typeof row.order_id === 'string' || row.order_id === null) &&
    typeof row.channel === 'string' &&
    typeof row.type === 'string' &&
    typeof row.status === 'string' &&
    typeof row.message_ru === 'string' &&
    typeof row.message_en === 'string' &&
    (typeof row.sent_at === 'string' || row.sent_at === null) &&
    typeof row.created_at === 'string'
  );
}

function assertNotificationPage(
  body: unknown,
): asserts body is NotificationFeedResponse {
  if (
    !body ||
    typeof body !== 'object' ||
    !Array.isArray((body as NotificationFeedResponse).notifications) ||
    !(body as NotificationFeedResponse).notifications.every(isNotificationItem) ||
    typeof (body as NotificationFeedResponse).total_count !== 'number' ||
    typeof (body as NotificationFeedResponse).page !== 'number' ||
    typeof (body as NotificationFeedResponse).per_page !== 'number'
  ) {
    throw new NotificationApiError(500, 'Invalid notification feed payload');
  }
}

// START_CONTRACT: listNotifications
//   PURPOSE: Fetch one page of the current customer's in-app notification feed.
//   INPUTS:  page: number, per_page: number.
//   OUTPUTS: Promise<NotificationFeedResponse>.
//   SIDE_EFFECTS: HTTP GET /api/v1/profile/notifications?page&per_page;
//                 throws NotificationApiError on non-2xx or schema mismatch.
//   LINKS:   PDD §5.2, INV-002, INV-013.
// END_CONTRACT: listNotifications
export async function listNotifications(
  page = 1,
  per_page = 20,
): Promise<NotificationFeedResponse> {
  const res = await authenticatedFetch(
    `/api/v1/profile/notifications?page=${page}&per_page=${per_page}`,
  );
  if (!res.ok) throw await parseError(res);
  const body = await res.json();
  assertNotificationPage(body);
  return body;
}

File: services/core-api/src/core_api/routers/notifications.py
"""Customer notification feed routes (PDD §5.2, INV-013)."""
from __future__ import annotations

# START_MODULE_CONTRACT
#   PURPOSE: HTTP route for the customer's own in-app notification feed under
#            /api/v1/profile/notifications.
#   SCOPE:   Read-only pagination over current user's notifications. CUSTOMER
#            only via rbac_matrix.ROUTE_MATRIX; no notification creation or
#            read-state mutation lives here.
#   DEPENDS: M-DATABASE (Session), core_api.deps.{auth,database},
#            core_api.services.notification_feed.
#   LINKS:   docs/development-plan.xml M-CORE-API, PDD §5.2, INV-002, INV-013.
#   ROLE:    RUNTIME
#   MAP_MODE: EXPORTS
# END_MODULE_CONTRACT
#
# START_MODULE_MAP
#   router                     - APIRouter("/api/v1/profile", tags=["notifications"])
#   get_my_notifications       - GET /api/v1/profile/notifications
# END_MODULE_MAP

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from core_api.deps import database as _db_dep
from core_api.deps.auth import get_current_user
from core_api.schemas.notification_feed import NotificationFeedResponse
from core_api.services.notification_feed import list_notifications_for_user

router = APIRouter(prefix="/api/v1/profile", tags=["notifications"])


def _get_session():
    yield from _db_dep.get_session()


# START_CONTRACT: get_my_notifications
#   PURPOSE: Paginated in-app notification feed for the current customer.
#   INPUTS:  page (1+), per_page (1..50), current_user, Session.
#   OUTPUTS: 200 NotificationFeedResponse.
#   SIDE_EFFECTS: none (read-only DB query).
#   LINKS:   PDD §5.2, INV-002, INV-013, services.notification_feed.
# END_CONTRACT: get_my_notifications
@router.get("/notifications", response_model=NotificationFeedResponse)
def get_my_notifications(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=50),
    current_user: dict = Depends(get_current_user),
    db: Session = Depends(_get_session),
) -> NotificationFeedResponse:
    return list_notifications_for_user(
        user_id=current_user["user_id"],
        page=page,
        per_page=per_page,
        db_session=db,
    )
